NBitcoin/Protocol/Node.cs

```diff
@@ -614,9 +614,12 @@ public static Node Connect(Network network, NodeConnectionParameters parameters
 				{
 					var timeout = new CancellationTokenSource(5000);
 					var param2 = parameters.Clone();
-					param2.ConnectCancellation = CancellationTokenSource.CreateLinkedTokenSource(parameters.ConnectCancellation, timeout.Token).Token;
-					var node = Node.Connect(network, addr.Endpoint, param2);
-					return node;
+					using(var cts = CancellationTokenSource.CreateLinkedTokenSource(parameters.ConnectCancellation, timeout.Token))
+					{
+						param2.ConnectCancellation = cts.Token;
+						var node = Node.Connect(network, addr.Endpoint, param2);
+						return node;
+					}

```

Comment:
We ar disposing `cts` after we are done with it. `Node.Connect` is not running asynchronous and that's why we are not disposing ahead of time.

Reply:
> Also why timeout not disposed here?

Because it is not leaking memory. It is garbage collected properly. The problem is not when we forget to dispose a cts but when we forget to dispose a `LinkedCancellationTockenSource` (lcts) that links a long-live cts (in this case `parameters.ConnectCancellation`) with a short-live cts (in this case `timeout.Token`). The GC cannot release the short-live cts because it is referenced by the lcts. This means that once we break the link by disposing the lcts, the GC releases the short-live cts as expected. Of course we can dispose it explicitly but it is not really necessary.

So, why aren't the lcts disposed by their finalizers and collected by the GC in the same way?  In general that's the case and the GC can release the lcts **except** when they allocates unmanaged resources (this is our case here). The lcts takes an unmanaged resource if we invoke the property `WaitHandle` (lazy loading it). In these cases we must dispose the lcts explicitly; otherwise we get a memory leak.

Reply:
> Of course we can dispose it explicitly but it is not really necessary.

I'd say, if something is IDisposable, let's dispose it. Don't trust* that there's a finalizer that does the job, just in case. This also makes the code less prone to be unfriendly to static analyzer (e.g. FxCop, Gendarme, etc). *-> verify!